x-pack/plugins/security/server/routes/analytics/record_violations.ts

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { TypeOf } from '@kbn/config-schema';
+import { schema } from '@kbn/config-schema';
+
+import type { RouteDefinitionParams } from '..';
+
+/**
+ * # Tracking CSP violations
+ *
+ * Add the following settings to your `kibana.dev.yml`:
+ *
+ * ```yml
+ * server.customResponseHeaders.Reporting-Endpoints: violations-endpoint="https://localhost:5601/xyz/internal/security/analytics/_record_violations"
+ * csp.report_to: [violations-endpoint]
+ * ```
+ *
+ * Note: The endpoint has to be on HTTPS and has to be a fully qualified URL including protocol and
+ * hostname, otherwise browsers might not send any reports. When using `server.publicBaseUrl` setting
+ * you should use the same origin for the reporting endpoint since Kibana does not support
+ * cross-origin content negotiation so browsers would not be able to send any report.
+ *
+ * # Debugging CSP violations
+ *
+ * CSP violations are tracked (batched) using event based telemetry.
+ *
+ * To print telemetry events to the terminal add the following config to your `kibana.dev.yml`:
+ *
+ * ```yml
+ * logging.loggers:
+ *   - name: analytics
+ *     level: all
+ *     appenders: [console]
+ * ```
+ */
+
+/**
+ * Schema that validates CSP violation reports according to W3C spec.
+ *
+ * https://www.w3.org/TR/CSP3/#reporting
+ */
+const cspViolationReportSchema = schema.object(
+  {
+    type: schema.literal('csp-violation'),
+    age: schema.maybe(schema.number()),
+    url: schema.string(),
+    user_agent: schema.maybe(schema.string()),
+    body: schema.object(
+      {
+        documentURL: schema.string(),
```

Comment:
Nit: suggestion to add inline description comments like for `permissionsPolicyViolationReportSchema` below.

Reply:
I couldn't find any useful definitions for these and I don't want to make them up which is why I linked to the w3 docs above